### PR TITLE
Refactoring text_editing.dart

### PR DESCRIPTION
**Pull request by nturgut**
Refactoring text_editing.dart to use EditingState more efficiently and removing unused/unimplemented code.

### CODE REVIEW

File: lib/web_ui/lib/src/engine/text_editing.dart [hunk 1]
```diff
@@ -66,6 +66,29 @@ class EditingState {
         baseOffset = flutterEditingState['selectionBase'],
         extentOffset = flutterEditingState['selectionExtent'];
 
+  /// Creates an [EditingState] instance using values from the editing element
+  /// in the dom.
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/text_editing.dart]: nit: DOM (all caps)

File: lib/web_ui/lib/src/engine/text_editing.dart [hunk 2]
```diff
@@ -66,6 +66,29 @@ class EditingState {
...
+  ///
+  /// This element can be a InputElement or a TextAreaElement depending on the
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/text_editing.dart]: nit: replace "This element" with "[domElement]"

Comment (yjbanov) [inline lib/web_ui/lib/src/engine/text_editing.dart]: nit: make "InputElement" and "TextAreaElement" linkable.

File: lib/web_ui/lib/src/engine/text_editing.dart [hunk 3]
```diff
@@ -66,6 +66,29 @@ class EditingState {
...
+  /// [InputType] of the text field.
+  factory EditingState.fromDomElement(html.HtmlElement domElement) {
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/text_editing.dart]: Why not call `configureDomElementFromEditingState`?

File: lib/web_ui/lib/src/engine/text_editing.dart [hunk 4]
```diff
@@ -110,6 +133,24 @@ class EditingState {
         ? 'EditingState("$text", base:$baseOffset, extent:$extentOffset)'
         : super.toString();
   }
+
+  /// Sets the selection values of a dom element using [EditingState].
+  ///
+  /// This element can be a InputElement or a TextAreaElement depending on the
+  /// [InputType] of the text field.
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/text_editing.dart]: same suggestion as above

File: lib/web_ui/lib/src/engine/text_editing.dart [hunk 5]
```diff
@@ -436,93 +423,23 @@ class TextEditingElement {
...
-  ///   that the element has been swapped.
-  void _swapDomElement(ElementType newElementType) {
-    // TODO(mdebbar): Create the appropriate dom element and initialize it.
-  }
-
   void _handleChange(html.Event event) {
-    _lastEditingState = calculateEditingState();
-    _onChange(_lastEditingState);
+    EditingState newEditingState = calculateEditingState();
+    if (newEditingState != _lastEditingState) {
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/text_editing.dart]: `EditingState` does not implement `operator ==`, so this only compares identity of two objects. Is that what we want to do here?

File: lib/web_ui/lib/src/engine/text_editing.dart (hunk 3 above)
Comment (mdebbar) [reply lib/web_ui/lib/src/engine/text_editing.dart]: `EditingState.fromDomElement` and `EditingState.configureDomElementFromEditingState` do opposite things. The former reads the state from the DOM element. The latter sets the state on the DOM element.

File: lib/web_ui/lib/src/engine/text_editing.dart (hunk 5 above)
Comment (mdebbar) [reply lib/web_ui/lib/src/engine/text_editing.dart]: It does implement `operator ==`.

Comment (yjbanov) [reply lib/web_ui/lib/src/engine/text_editing.dart]: Ah, nvm then. I got confused by Github UI and didn't find it.

File: lib/web_ui/lib/src/engine/text_editing.dart [hunk 6]
```diff
@@ -110,6 +133,24 @@ class EditingState {
         ? 'EditingState("$text", base:$baseOffset, extent:$extentOffset)'
         : super.toString();
   }
+
+  /// Sets the selection values of a dom element using [EditingState].
+  ///
+  /// This element can be a InputElement or a TextAreaElement depending on the
+  /// [InputType] of the text field.
+  void configureDomElementFromEditingState(html.HtmlElement domElement) {
```
[the remark this answers is not in the document]
Comment (nturgut) [reply lib/web_ui/lib/src/engine/text_editing.dart]: Done.

File: lib/web_ui/lib/src/engine/text_editing.dart [hunk 7]
```diff
@@ -110,6 +133,24 @@ class EditingState {
...
+    if (domElement is html.InputElement) {
+      html.InputElement element = domElement;
+      element.value = text;
+      element.setSelectionRange(baseOffset, extentOffset);
+    } else if (domElement is html.TextAreaElement) {
+      html.TextAreaElement element = domElement;
+      element.value = text;
+      element.setSelectionRange(baseOffset, extentOffset);
+    } else {
+      throw UnsupportedError('Initialized with unsupported input type');
```
[the remark this answers is not in the document]
Comment (nturgut) [reply lib/web_ui/lib/src/engine/text_editing.dart]: Done.

File: lib/web_ui/lib/src/engine/text_editing.dart [hunk 8]
```diff
@@ -309,15 +314,24 @@ class TextEditingElement {
       ..add(html.document.onSelectionChange.listen(_handleChange))
```
[the remark this answers is not in the document]
Comment (nturgut) [reply lib/web_ui/lib/src/engine/text_editing.dart]: Sure! Thanks!

File: lib/web_ui/lib/src/engine/text_editing.dart [hunk 9]
```diff
@@ -163,33 +204,6 @@ class InputConfiguration {
 
 typedef _OnChangeCallback = void Function(EditingState editingState);
 
-enum ElementType {
```
[the remark this answers is not in the document]
Comment (nturgut) [reply lib/web_ui/lib/src/engine/text_editing.dart]: Np!

File: lib/web_ui/lib/src/engine/text_editing.dart (hunk 5 above)
[the remark this answers is not in the document]
Comment (nturgut) [reply lib/web_ui/lib/src/engine/text_editing.dart]: When Flutter calls setEditingState, it immediately sets the _lastEditingState.

File: lib/web_ui/lib/src/engine/text_editing.dart [hunk 10]
```diff
@@ -309,15 +314,24 @@ class TextEditingElement {
...
-    // See [SelectionChangeDetection].
+    /// Detects changes in text selection.
+    ///
+    /// Currently only used in Firefox.
+    ///
+    /// In Firefox, when cursor moves, neither selectionChange nor onInput
+    /// events are triggered. We are listening to keyup event. Selection start,
+    /// end values are used to decide if the text cursor moved.
+    ///
+    /// Specific keycodes are not checked since users/applicatins can bind their own
```
[the remark this answers is not in the document]
Comment (nturgut) [reply lib/web_ui/lib/src/engine/text_editing.dart]: Done.

File: lib/web_ui/lib/src/engine/text_editing.dart [hunk 11]
```diff
@@ -436,93 +423,23 @@ class TextEditingElement {
...
-    _onChange(_lastEditingState);
+    EditingState newEditingState = calculateEditingState();
+    if (newEditingState != _lastEditingState) {
+      _lastEditingState = newEditingState;
+      _onChange(_lastEditingState);
+    }
   }
 
   @visibleForTesting
   EditingState calculateEditingState() {
```
[the remark this answers is not in the document]
Comment (nturgut) [reply lib/web_ui/lib/src/engine/text_editing.dart]: done.

File: lib/web_ui/lib/src/engine/text_editing.dart [hunk 12]
```diff
@@ -550,10 +467,14 @@ class PersistentTextEditingElement extends TextEditingElement {
...
-    assert(_getTypeFromElement(domElement) != null);
+    assert((domElement is html.InputElement) ||
+        (domElement is html.TextAreaElement));
     this.domElement = domElement;
   }
 
+  // Unfortunately, in persistent mode, the user of this class has to be
+  // notified that the element is being swapped.
+  // TODO(mdebbar): do we need to call `old.replaceWith(new)` here?
   final html.VoidCallback _onDomElementSwap;
```
[the remark this answers is not in the document]
Comment (nturgut) [reply lib/web_ui/lib/src/engine/text_editing.dart]: done.

File: lib/web_ui/lib/src/engine/text_editing.dart (hunk 1 above)
Comment (nturgut) [reply lib/web_ui/lib/src/engine/text_editing.dart]: done.

File: lib/web_ui/lib/src/engine/text_editing.dart (hunk 2 above)
Comment (nturgut) [reply lib/web_ui/lib/src/engine/text_editing.dart]: done.

File: lib/web_ui/lib/src/engine/text_editing.dart (hunk 3 above)
Comment (nturgut) [reply lib/web_ui/lib/src/engine/text_editing.dart]: ack.

File: lib/web_ui/lib/src/engine/text_editing.dart (hunk 4 above)
Comment (nturgut) [reply lib/web_ui/lib/src/engine/text_editing.dart]: done.

File: lib/web_ui/lib/src/engine/text_editing.dart (hunk 5 above)
Comment (nturgut) [reply lib/web_ui/lib/src/engine/text_editing.dart]: ack.

File: lib/web_ui/lib/src/engine/text_editing.dart (hunk 2 above)
Comment (nturgut) [reply lib/web_ui/lib/src/engine/text_editing.dart]: done.

File: lib/web_ui/lib/src/engine/text_editing.dart [hunk 13]
```diff
@@ -305,20 +310,29 @@ class TextEditingElement {
...
+    ///
+    /// Specific keycodes are not checked since users/applications can bind
+    /// their own keys to move the text cursor.
+    /// Decides if the selection has changed (cursor moved) compared to the
+    /// previous values.
+    ///
+    /// After each keyup, the start/end values of the selection is compared to the
+    /// previously saved editing state.
     if (browserEngine == BrowserEngine.firefox) {
       _subscriptions.add(domElement.onKeyUp.listen((event) {
```
Comment (mdebbar) [inline lib/web_ui/lib/src/engine/text_editing.dart]: You also mentioned `domElement.onSelect` to fix the select all problem. If that’s coming in a separate PR, then nvm.

Comment (nturgut) [reply lib/web_ui/lib/src/engine/text_editing.dart]: yes it is coming on a separate PR. I want to make sure all browsers are covered for "Select All" context menu item.

**Pull request merged**